engine.go

```diff
@@ -0,0 +1,247 @@
+	tempDir, dockerTempDir, err := createTempDir()
+	if err != nil {
+		return "", "", err
+	}
+	err = createConfigurationFile(tempDir, lang)
+	if err != nil {
+		return "", "", err
+	}
+	err = createExecFile(tempDir, lang, codeBase)
+	if err != nil {
+		return "", "", err
+	}
+	return tempDir, dockerTempDir, nil
+}
+
+func prepareAndSimpleRun(w http.ResponseWriter, r *http.Request, tempDir, dockerTempDir, codeBase string) {
+	key, build := piper.LogBuildStart("challengeId", codeBase, "user")
+
+	cmdUser := exec.Command(
+		"docker",
+		"run",
+		"--rm",
+		"-v",
+		dockerTempDir+":/run",
+		"coduno_all")
+
+	outUser, err := cmdUser.StdoutPipe()
+	if err != nil {
+		log.Fatal(err)
+	}
+	errUser, err := cmdUser.StderrPipe()
+	if err != nil {
+		log.Fatal(err)
+	}
+	_, err = cmdUser.StdinPipe()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var runOutput, runErr bytes.Buffer
+	cmdUser.Start()
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go piper.PipeOutput(&wg, outUser, os.Stdout, &runOutput)
+	go piper.PipeOutput(&wg, errUser, os.Stdout, &runErr)
+
+	exitErr := cmdUser.Wait()
+	wg.Wait()
+	prepLog, err := ioutil.ReadFile(tempDir + "/prepare.log")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var stats syscall.Rusage
+	statsData, err := ioutil.ReadFile(tempDir + "/stats.log")
+	if err != nil {
+		log.Print(err)
+	} else {
+		err = json.Unmarshal(statsData, &stats)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	piper.LogRunComplete(key, build, "", runOutput.String(), "", exitErr, string(prepLog), stats)
+
+	var toSend = make(map[string]string)
+	toSend["run"] = runOutput.String()
+	toSend["err"] = runErr.String()
+
+	json, err := json.Marshal(toSend)
+	if err != nil {
+		http.Error(w, "Json marshal err: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.Write(json)
+}
+
+func createExecFile(tmpDir, lang, codeBase string) error {
+	f, err := os.Create(tmpDir + "/" + fileNames[lang])
+	if err != nil {
+		return err
+	}
+	f.WriteString(codeBase)
+	f.Close()
+	return nil
+}
+
+func createConfigurationFile(tempDir, lang string) error {
+	if err := copyFileContents(languages[lang], tempDir, configFileName); err != nil {
+		return err
+	}
+	return nil
+}
+
+func createTempDir() (string, string, error) {
+	id, err := uuid.V4()
+	if err != nil {
+		return "", "", err
+	}
+	tempDir := homeDir + id.String()
+	os.MkdirAll(tempDir, 0777)
+
+	return tempDir, dockerHomeDir + id.String(), nil
+}
+
+// copyFileContents copies the contents of the file named src to the file named
+// by dst. The file will be created if it does not already exist. If the
+// destination file exists, all it's contents will be replaced by the contents
+// of the source file.
+func copyFileContents(src, dst, fileName string) (err error) {
+	in, err := os.Open(src)
+	if err != nil {
+		return
+	}
+	dst = dst + "/" + fileName
+	defer in.Close()
+	out, err := os.Create(dst)
+	if err != nil {
+		return
+	}
+	defer func() {
+		cerr := out.Close()
+		if err == nil {
+			err = cerr
```

Comment:
This will set `err` to `nil` if `out.Close()` succeeds, effectively overwriting previous errors. Is that intended?